src/main/java/com/opentable/components/filterorder/CycleFinder.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.opentable.components.filterorder;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Given a mapping of graph nodes to other nodes (&ldquo;edges&rdquo;), call {@link #run} to find a cycle, if there
+ * is one.
+ * @param <T> node type
+ */
+class CycleFinder<T> {
+    private final Set<T> seen = new HashSet<>();
+    private final Map<T, Collection<T>> edges;
+
+    CycleFinder(final Map<T, Collection<T>> edges) {
+        this.edges = edges;
+    }
+
+    /** Return a list of cyclic nodes, if there is one, otherwise empty. */
+    Optional<List<T>> run() {
```

Comment:
Why optional of list? Just use an empty list to indicate no cycle.

Reply:
We could do that.  My thinking was that an empty optional was one further step to indicate that there _really_ is nothing for you to look at.  It prevents any ambiguity.